src/resolve/resolved-dns-netservice.c

```diff
@@ -157,3 +157,97 @@ void dns_netservice_remove_all(DnsNetservice *first) {
+
+        next = first->netservices_next;
+
+        if (first->ptr_rr)
+                first->ptr_rr = dns_resource_record_unref(first->ptr_rr);
+        if (first->srv_rr)
+                first->srv_rr = dns_resource_record_unref(first->srv_rr);
+        if (first->txt_rr)
+                first->txt_rr = dns_resource_record_unref(first->txt_rr);
+
+        service_name = strjoina(first->type, ".local");
+        instance_name = strjoina(first->instance_name, ".", first->type, ".local");
+
+        first->txt_rr = dns_resource_record_new_full(DNS_CLASS_IN, DNS_TYPE_TXT,
+                                                     instance_name);
+        if (!first->txt_rr)
+                goto oom;
+
+        if (strv_length(first->txt) > 0) {
+                DnsTxtItem *last = NULL;
+                char **value;
+
+                STRV_FOREACH(value, first->txt) {
+                        DnsTxtItem *i;
+                        size_t sz;
+
+                        sz = strlen(*value);
+                        i = malloc0(offsetof(DnsTxtItem, data) + sz + 1);
+                        if (!i)
+                                goto oom;
+
+                        memcpy(i->data, *value, sz);
+                        i->length = sz;
+
+                        LIST_INSERT_AFTER(items, first->txt_rr->txt.items, last, i);
+                        last = i;
+                }
+        } else {
+                DnsTxtItem *i;
+
+                /* RFC 6763, section 6.1 suggests to treat
+                 * empty TXT RRs as equivalent to a TXT record
+                 * with a single empty string. */
+
+                i = malloc0(offsetof(DnsTxtItem, data) + 1); /* for safety reasons we add an extra NUL byte */
+                if (!i)
+                        goto oom;
+
+                first->txt_rr->txt.items = i;
+        }
+        first->txt_rr->ttl = MDNS_DEFAULT_TTL;
+
+        first->ptr_rr = dns_resource_record_new_full(DNS_CLASS_IN, DNS_TYPE_PTR,
+                                                     service_name);
+        if (!first->ptr_rr)
+                goto oom;
+
+        first->ptr_rr->ttl = MDNS_DEFAULT_TTL;
+        first->ptr_rr->ptr.name = strdup(instance_name);
+        if (!first->ptr_rr->ptr.name)
+                goto oom;
+
+        first->srv_rr = dns_resource_record_new_full(DNS_CLASS_IN, DNS_TYPE_SRV,
+                                                     instance_name);
+        if (!first->srv_rr)
+                goto oom;
+
+        first->srv_rr->ttl = MDNS_DEFAULT_TTL;
+        first->srv_rr->srv.priority = first->priority;
+        first->srv_rr->srv.weight = first->weight;
+        first->srv_rr->srv.port = first->port;
+        first->srv_rr->srv.name = strdup(hostname);
+        if (!first->srv_rr->srv.name)
+                goto oom;
+
+        return dns_netservice_update_rrs(next, hostname);
+
+oom:
+        if (first->txt_rr)
+                first->txt_rr = dns_resource_record_unref(first->txt_rr);
```

Comment:
our _unref() functions happily accept NULL as input and become NOPs then, hence the explicit if check is unnecessary.